rust/kernel/device.rs

```diff
@@ -0,0 +1,59 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Generic devices that are part of the kernel's driver model.
+//!
+//! C header: [`include/linux/device.h`](../../../../include/linux/device.h)
+
+use crate::bindings;
+
+/// A raw device.
+///
+/// # Safety
+///
+/// Implementers must ensure that the `*mut device` returned by [`RawDevice::raw_device`] is
+/// related to `self`, that is, actions on it will affect `self`. For example, if one calls
+/// `get_device`, then the refcount on the device represented by `self` will be incremented.
+pub unsafe trait RawDevice {
+    /// Returns the raw `struct device` related to `self`.
+    fn raw_device(&self) -> *mut bindings::device;
+}
+
+/// A ref-counted device.
+///
+/// # Invariants
+///
+/// `ptr` is valid, non-null, and has a non-zero reference count. One of the references is owned by
+/// `self`, and will be decremented when `self` is dropped.
+pub struct Device {
```

Comment:
Maybe call it `AnyDevice` to disambiguate from the driver specific `Device` type?

Reply:
My intent here is to use the module name to differentiate these, for example `amba::Device`, `pci::Device`, etc. This is similar to other names like `Chip`, where we have `gpio::Chip`, `irq::Chip`, etc. And then if people want use `AnyDevice`, for example, they can use `use device::Device as AnyDevice` and variants to bind different names.

Does this sound weird or unorthodox?

Reply:
I guess that would work.

Reply:
There is also the advantage of using the same name as in the C side.